crypto/dh/dh_gen.c

```diff
@@ -57,8 +57,12 @@ int dh_generate_ffc_parameters(DH *dh, int type, int pbits,
         ret = ffc_params_FIPS186_4_generate(dh->libctx, &dh->params,
                                             FFC_PARAM_TYPE_DH,
                                             pbits, qbits, md, &res, cb);
-    if (ret > 0)
+    if (ret > 0) {
         dh->dirty_cnt++;
+    } else {
+        if ((res & FFC_CHECK_BAD_LN_PAIR) != 0)
+            DHerr(0, DH_R_BAD_FFC_PARAMETERS);
+    }
```

Comment:
Why not add errors in `ffc_params_FIPS186_2_generate` and `ffc_params_FIPS186_4_generate`?

Reply:
> Why not add errors in ffc_params_FIPS186_2_generate and ffc_params_FIPS186_4_generate?

This would have been my preferred solution, but so far there have been no `FFCerr()` calls and `FFC_R_*` codes. 
Shall I introduce those, or move the `DHerr(0, DH_R_BAD_FFC_PARAMETERS)` and `DSAerr(0, DSA_R_BAD_FFC_PARAMETERS)` calls into the appropriate places?

BTW, as mentioned in https://github.com/openssl/openssl/issues/12047#issuecomment-638967154 the best place for the err() calls in my view would be directly in `ffc_params_generate.c`, for instance:
```
static int ffc_validate_LN(size_t L, size_t N, int type)
{
    if (type == FFC_PARAM_TYPE_DH) {
        /* Valid DH L,N parameters from SP800-56Ar3 5.5.1 Table 1 */
        if (L == 2048 && (N == 224 || N == 256))
            return 112;
#ifndef OPENSSL_NO_DH
        DHerr(0, DH_R_BAD_FFC_PARAMETERS);
#endif
    } else if (type == FFC_PARAM_TYPE_DSA) {
        /* Valid DSA L,N parameters from FIPS 186-4 Section 4.2 */
        if (L == 1024 && N == 160)
            return 80;
        if (L == 2048 && (N == 224 || N == 256))
            return 112;
        if (L == 3072 && N == 256)
            return 128;
#ifndef OPENSSL_NO_DSA
        DSAerr(0, DSA_R_BAD_FFC_PARAMETERS);
#endif
    }
    return 0;
}
```

Reply:
I've tentatively moved the error queue entry generation there.
Or should it actually better be in `ffc_params_FIPS186_2_generate` and `ffc_params_FIPS186_4_generate`?

Reply:
We could also unify the names `DSA_R_BAD_FFC_PARAMETERS` and `DH_R_BAD_FFC_PARAMETERS` - if so, to which one?

Reply:
I would suggest not using WHATEVERerr(), but use ERR_raise() or ERR_raise_data() instead.  There's much less to deal with there.

I leave it to you to use ERR_LIB_DSA and ERR_LIB_DH, or create ERR_LIB_FFC and use that if you prefer.

Reply:
Personally, I'm perfectly OK with using `DSA_R_BAD_FFC_PARAMETERS` for the DSA stuff and `DH_R_BAD_FFC_PARAMETERS` for the DH stuff, and am OK with the error messages in `ffc_validate_LN`

Reply:
Actually, I just looked one more time, and am happy with what you did